File: engine/src/com/magma/engine/chars/CharacterController.java
package com.magma.engine.chars;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input.Keys;
import com.badlogic.gdx.scenes.scene2d.Action;
import com.magma.engine.entities.Entity;
import com.magma.engine.input.GameInput;
import com.magma.engine.ui.dialog.Dialog;
import com.magma.engine.utils.Time;

public class CharacterController extends Action {

	public static final int key_up = GameInput.registerKey("move_up", Keys.W); // 0
	public static final int key_down = GameInput.registerKey("move_down", Keys.S); // 1
	public static final int key_left = GameInput.registerKey("move_left", Keys.A); // 2
	public static final int key_right = GameInput.registerKey("move_right", Keys.D); // 3

	protected Entity entity;
	protected boolean moveDiagonal;
	
	private boolean isMoving;
	private long[] pressTimes = new long[4];

	public enum Direction {
		Up, Down, Left, Right, None
	}

	/**
	 * Makes an actor movable with the keyboard. Default keys are WASD but can be
	 * changed in keybindings.json Designed for attaching to Characters
	 * 
	 * @param entity   Entity possessing the character
	 * @param diagonal Can the entity move in 8 directions (diagonal) or just 4.
	 */
	public CharacterController(Entity entity, boolean diagonal) {
		this.entity = entity;
		this.moveDiagonal = diagonal;
	}

	/**
	 * Makes an actor movable with the keyboard. Default keys are WASD but can be
	 * changed in keybindings.json Designed for attaching to Characters
	 * 
	 * @param entity Entity possessing the character
	 */
	public CharacterController(Entity entity) {
		this.entity = entity;
		this.moveDiagonal = false;
	}

	/**
	 * Makes an actor movable with the keyboard. Default keys are WASD but can be
	 * changed in keybindings.json Designed for attaching to Characters
	 * 
	 * @param speed Movement speed in pixels per second
	 */
	public CharacterController(float speed, boolean diagonal) {
		this.moveDiagonal = diagonal;
	}

	@Override
	public boolean act(float delta) {

        // stop the player from moving when dialog is playing
        if (Dialog.isSpeaking()){
            return false;
        }

		float sp = entity.getStats().speed * delta;

		// The character will move in the direction of the youngest pressed key.
		
		int[] allKeys = new int[] { key_up, key_down, key_left, key_right };
		
		for (int i = 0; i < allKeys.length; i++) {
			// pressing keys
			if (Gdx.input.isKeyJustPressed(allKeys[i])) {
				pressTimes[i] = Time.getTimeMillis();
                //MagmaLogger.log("Pressed",allKeys[i]);
			}

			//releasing
			if (!Gdx.input.isKeyPressed(allKeys[i])) {
				pressTimes[i] = Long.MIN_VALUE;
			}
		}

		// get the most recent pressed key of allKeys[]
		int largest = 0;
		for (int i = 0; i < allKeys.length; i++) {
			if (pressTimes[i] >= pressTimes[largest]) {
				largest = i;
			}
		}
	
		
		if (pressTimes[largest] > 0) {
			isMoving = true;
			switch (largest) {
			case 0: //key_up
				getActor().moveBy(0, sp);
				break;
			case 1: //key_down
				getActor().moveBy(0, -sp);
				break;
			case 2: //key_left
				getActor().moveBy(-sp, 0);
				break;
			case 3: //key_right
				getActor().moveBy(sp, 0);
				break;
			default: // none
				break;
			}
		}else {
			// no key being pressed
			isMoving = true;
		}
		return false;
	}

	public boolean isMoving() {
		return isMoving;
	}
	
	@Override
	public String toString() {
		return "CharacterController sp:" + entity.getStats().speed + " moving? " + isMoving;
	}
}

File: engine/src/com/magma/engine/chars/CharacterController.kt
package com.magma.engine.chars

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input
import com.badlogic.gdx.scenes.scene2d.Action
import com.magma.engine.entities.Entity
import com.magma.engine.input.GameInput
import com.magma.engine.ui.dialog.Dialog
import com.magma.engine.utils.Time
import java.lang.IllegalArgumentException

/**
 * Makes an actor movable with the keyboard. Default keys are WASD but can be
 * changed in keybindings.json Designed for attaching to Characters
 *
 * @param entity   Entity possessing the character
 * @param diagonal Can the entity move in 8 directions (diagonal) or just 4.
 */
class CharacterController(private val entity: Entity, private val diagonal: Boolean = false) : Action() {

    var isMoving = false
        private set

    private val pressTimes = LongArray(4)

    enum class Direction {
        Up, Down, Left, Right, None
    }

    override fun act(delta: Float): Boolean {
        // stop the player from moving when dialog is playing
        if (Dialog.isSpeaking()) {
            return false
        }
        val sp = entity.stats.speed * delta

        // The character will move in the direction of the youngest pressed key.
        val allKeys = intArrayOf(key_up, key_down, key_left, key_right)
        for (i in allKeys.indices) {
            // pressing keys
            if (Gdx.input.isKeyJustPressed(allKeys[i])) {
                pressTimes[i] = Time.getTimeMillis()
                //MagmaLogger.log("Pressed",allKeys[i]);
            }

            //releasing
            if (!Gdx.input.isKeyPressed(allKeys[i])) {
                pressTimes[i] = Long.MIN_VALUE
            }
        }

        // get the most recent pressed key of allKeys[]
        var largest = 0
        for (i in allKeys.indices) {
            if (pressTimes[i] >= pressTimes[largest]) {
                largest = i
            }
        }
        if (pressTimes[largest] > 0) {
            isMoving = true
            when (largest) {
                0 -> getActor().moveBy(0f, sp)
                1 -> getActor().moveBy(0f, -sp)
                2 -> getActor().moveBy(-sp, 0f)
                3 -> getActor().moveBy(sp, 0f)
                else -> { throw IllegalArgumentException("Unknown direction")}
            }
        } else {
            // no key being pressed
            isMoving = true
        }
        return false
    }

    override fun toString(): String {
        return "CharacterController sp:" + entity.stats.speed + " moving? " + isMoving
    }

    companion object {
        val key_up = GameInput.registerKey("move_up", Input.Keys.W) // 0
        val key_down = GameInput.registerKey("move_down", Input.Keys.S) // 1
        val key_left = GameInput.registerKey("move_left", Input.Keys.A) // 2
        val key_right = GameInput.registerKey("move_right", Input.Keys.D) // 3
    }
}